company.el

```diff
@@ -1187,14 +1187,19 @@ can retrieve meta-data for them."
             (cl-dotimes (i (1+ len))
               (when (setq prev (cdr (assoc (substring prefix 0 (- len i))
                                            company-candidates-cache)))
-                (setq candidates (all-completions prefix prev))
-                (cl-return t)))))
+                (cl-return t)))
+            (cond ((eq prev 'async)
+                   ;; Pretend we found results so we don't trigger another search.
+                   t)
+                  (prev
+                   (setq candidates (all-completions prefix prev))))))
         (progn
           ;; No cache match, call the backend.
           (setq candidates (company--preprocess-candidates
                             (company--fetch-candidates prefix)))
           ;; Save in cache.
-          (push (cons prefix candidates) company-candidates-cache)))
+          (if candidates
+              (push (cons prefix candidates) company-candidates-cache))))
```

Comment:
No particular reason. It's just simpler without the condition, and the async handler overwrites the cache later anyway.